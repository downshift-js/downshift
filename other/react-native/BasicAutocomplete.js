import React from 'react'
import {Text, TextInput, TouchableOpacity, View} from 'react-native'
import Downshift from '../../dist/downshift.native.cjs'

const MyView = ({innerRef, ...rest}) => <View ref={innerRef} {...rest} />

const BasicAutocomplete = ({items, ...rest}) => (
  <Downshift
    {...rest}
    render={({
      getRootProps,
      getInputProps,
      getItemProps,
      isOpen,
      inputValue,
      selectedItem,
      highlightedIndex,
    }) => (
      <MyView {...getRootProps({refKey: 'innerRef'})}>
        <TextInput {...getInputProps({placeholder: 'Favorite color ?'})} />
        {isOpen ? (
          <View style={{borderWidth: 1, borderColor: '#ccc'}}>
            {items
              .filter(
                i =>
                  !inputValue ||
                  i.toLowerCase().includes(inputValue.toLowerCase()),
              )
              .map((item, index) => {
                const props = getItemProps({item, index})
                return (
                  <TouchableOpacity {...props} key={item}>
                    <View
                      style={{
                        backgroundColor:
                          highlightedIndex === index ? 'gray' : 'white',
                      }}
                    >
                      <Text
                        style={{
                          fontWeight: selectedItem === item ? 'bold' : 'normal',
                        }}
                      >
                        {item}
                      </Text>
                    </View>
                  </TouchableOpacity>
                )
              })}
          </View>
        ) : null}
      </MyView>
    )}
  />
)

export default BasicAutocomplete

/*
 eslint
  react/prop-types: 0,
  import/extensions: 0,
  import/no-unresolved: 0
 */
