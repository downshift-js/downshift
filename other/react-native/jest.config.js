const jestConfig = require('kcd-scripts/config').jest

module.exports = Object.assign(jestConfig, {
  preset: 'react-native',
  testEnvironment: 'node',
  transformIgnorePatterns: ['<rootDir>/node_modules/(?!react-native)/'],
  rootDir: '../../',
  roots: ['.'],
  testMatch: ['<rootDir>/other/react-native/__tests__/**/*.js?(x)'],
})
