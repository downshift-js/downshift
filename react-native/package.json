{
  "main": "dist/downshift.cjs.js",
  "jsnext:main": "dist/downshift.esm.js",
  "module": "dist/downshift.esm.js"
}
