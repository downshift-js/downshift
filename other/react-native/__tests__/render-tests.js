/* eslint-disable react/prop-types */
// eslint-disable-next-line import/no-unassigned-import
import {Text, TextInput, View} from 'react-native'
import React from 'react'

// Note: test renderer must be required after react-native.
import TestRenderer from 'react-test-renderer'

import Downshift from '../../../dist/downshift.native.cjs'

test('renders with React Native components', () => {
  const RootView = ({innerRef, ...rest}) => <View ref={innerRef} {...rest} />
  const renderSpy = jest.fn(({getRootProps, getInputProps, getItemProps}) => (
    <RootView {...getRootProps({refKey: 'innerRef'})}>
      <TextInput {...getInputProps()} />
      <View>
        <Text {...getItemProps({item: 'foo', index: 0})}>foo</Text>
        <Text {...getItemProps({item: 'bar', index: 1})}>bar</Text>
      </View>
    </RootView>
  ))
  const element = <Downshift render={renderSpy} />
  const renderer = TestRenderer.create(element)
  expect(renderSpy).toHaveBeenCalledWith(
    expect.objectContaining({
      isOpen: false,
      highlightedIndex: null,
      selectedItem: null,
      inputValue: '',
    }),
  )
  const tree = renderer.toJSON()
  expect(tree).toMatchSnapshot()
})

test('can use children instead of render prop', () => {
  const RootView = ({innerRef, ...rest}) => <View ref={innerRef} {...rest} />
  const renderSpy = jest.fn(({getRootProps, getInputProps, getItemProps}) => (
    <RootView {...getRootProps({refKey: 'innerRef'})}>
      <TextInput {...getInputProps()} />
      <View>
        <Text {...getItemProps({item: 'foo', index: 0})}>foo</Text>
        <Text {...getItemProps({item: 'bar', index: 1})}>bar</Text>
      </View>
    </RootView>
  ))
  const element = <Downshift>{renderSpy}</Downshift>
  TestRenderer.create(element)
  expect(renderSpy).toHaveBeenCalledTimes(1)
})

test('calls onChange when TextInput changes values', () => {
  const onChange = jest.fn()
  const Input = jest.fn(props => <TextInput {...props} />)

  const RootView = ({innerRef, ...rest}) => <View ref={innerRef} {...rest} />
  const renderSpy = jest.fn(({getRootProps, getInputProps, getItemProps}) => (
    <RootView {...getRootProps({refKey: 'innerRef'})}>
      <Input {...getInputProps({onChange})} />
      <View>
        <Text {...getItemProps({item: 'foo', index: 0})}>foo</Text>
        <Text {...getItemProps({item: 'bar', index: 1})}>bar</Text>
      </View>
    </RootView>
  ))
  const element = <Downshift>{renderSpy}</Downshift>
  TestRenderer.create(element)
  expect(renderSpy).toHaveBeenCalledTimes(1)

  const [[firstArg]] = Input.mock.calls
  expect(firstArg).toMatchObject({
    // TODO: We shouldn't need to know about the internals of how we're affecting the TextInput and what props we're supplying.
    // See https://github.com/paypal/downshift/issues/361
    onChangeText: expect.any(Function),
  })
  const fakeEvent = 'foobar'
  firstArg.onChangeText(fakeEvent)

  expect(onChange).toHaveBeenCalledTimes(1)
  expect(onChange).toHaveBeenCalledWith(fakeEvent)
})

/*
 eslint
  react/prop-types: 0,
  import/extensions: 0,
  import/no-unresolved: 0
 */
