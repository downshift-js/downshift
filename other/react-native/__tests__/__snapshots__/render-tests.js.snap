// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with React Native components 1`] = `
<View
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-labelledby="downshift-0-label"
  aria-owns={null}
  role="combobox"
>
  <TextInput
    allowFontScaling={true}
    aria-activedescendant={null}
    aria-autocomplete="list"
    aria-controls={null}
    aria-labelledby="downshift-0-label"
    autoComplete="off"
    id="downshift-0-input"
    onBlur={[Function]}
    onChangeText={[Function]}
    onKeyDown={[Function]}
    value=""
  />
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      aria-selected={false}
      ellipsizeMode="tail"
      id="downshift-0-item-0"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onPress={[Function]}
      role="option"
    >
      foo
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      aria-selected={false}
      ellipsizeMode="tail"
      id="downshift-0-item-1"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onPress={[Function]}
      role="option"
    >
      bar
    </Text>
  </View>
</View>
`;
